Initial refactoring for completion suggester

areek opened the pull request:
This PR implements serialization methods for CompletionSuggestionBuilder based on the suggest refactoring work started in https://github.com/elastic/elasticsearch/pull/16241 and improves test coverage. This simplifies CompletionSuggestionBuilder by making fuzzy, regex and context queries implement their own serialization methods.

cbuescher commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java:
```diff
@@ -374,25 +182,83 @@ public String getWriteableName() {
+                out.writeString(payloadField);
+            }
+        } else {
+            out.writeBoolean(false);
```
nit: Everything okay, but I found this (and the following same constructs) a little hard to read. Can I suggest writing the boolean flag first without conditions (e.g. `out.writeBoolean(fuzzyOptions != null)` and then have the if-block only? But either was is fine.

areek replied:
Now we write the boolean flag first followed by an if-only block

cbuescher commented on core/src/main/java/org/elasticsearch/search/suggest/completion/context/CategoryQueryContext.java:
```diff
@@ -118,7 +110,8 @@ public CategoryQueryContext build() {
         CATEGORY_PARSER.declareBoolean(Builder::setPrefix, new ParseField("prefix"));
```
Can these ParseFields be linked to the String constants in CategoryContextMapping?

areek replied:
Oops, great catch. linked the ParseFields in GeoContextMapping as well

cbuescher commented on core/src/main/java/org/elasticsearch/search/suggest/completion/context/GeoContextMapping.java:
```diff
@@ -247,18 +247,15 @@ protected XContentBuilder toInnerXContent(XContentBuilder builder, Params params
-            int minPrecision = this.precision;
-            if (queryContext.getPrecision() != -1) {
-                minPrecision = Math.min(minPrecision, queryContext.getPrecision());
-            }
```
Is this deletion related to the refactoring or just something to do in general? Just asking because I don't get the details here.

areek replied:
This is due to the change in the default `precision` for `GeoQueryContext.Builder` see [here](https://github.com/elastic/elasticsearch/pull/16398/files#diff-6252417c21bed452f272fe142129fed7R192). before `-1` indicated the index-time precision value be used for query contexts, now we set the default to `12` (highest precision) and take the minimum precision between the index and query time precision.

cbuescher commented on core/src/test/java/org/elasticsearch/search/suggest/AbstractSuggestionBuilderTestCase.java:
```diff
@@ -166,7 +168,8 @@ protected SB serializedCopy(SB original) throws IOException {
     }
 
-    protected static <T> void maybeSet(Consumer<T> consumer, T value) {
+    // TODO: move this to ESTestCase?
```
+1 but maybe this helper so far is limited to the query / suggester tests only. But I think its general enough.

areek replied:
I moved `maybeSet` and `randomValueOtherThan` to `ESTestCase`

cbuescher commented on core/src/test/java/org/elasticsearch/search/suggest/completion/CategoryQueryContextTests.java:
```diff
@@ -0,0 +1,99 @@
+        builder.setCategory(original.getCategory()).setBoost(original.getBoost()).setPrefix(original.isPrefix());
+        switch (randomIntBetween(0, 2)) {
+            case 0:
+                builder.setCategory(original.getCategory() + randomAsciiOfLength(10));
```
Maybe take a look if you want to use AbstractSuggestionBuilderTestCase#randomValueOtherThan() here. That one could probably also go into a separate test utility class, wdyt?

areek replied:
Thanks for the suggestion!

cbuescher commented on core/src/test/java/org/elasticsearch/search/suggest/completion/WritableTestCase.java:
```diff
@@ -0,0 +1,104 @@
+ * Base class for testing serialization and equality for
+ * {@link Writeable} models
+ */
+public abstract class WritableTestCase<M extends Writeable> extends ESTestCase {
```
+1 on pulling this out, I'm sure this can be used in other places, although in many of the tests I'm thinking about right now we have NamedWriteable things under test, but only the copy method would need to be slightly different.

cbuescher commented:
@areek looks great, I left a few small comments but other than that tests look good and also pulling out all those separate classes makes things much easier to understand I think.

areek commented:
@cbuescher Thanks for the review. I addressed all your comments. 
`CompletionSuggestionBuilder#innerFromXContent` has not been implemented yet. We should clean up the parser in `CompletionSuggestParser` to use CompletionSuggestionBuilder as the consumer and fix setters with multiple params [(prefix, fuzzy), (prefix, regex), etc.] before.

cbuescher commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -126,6 +128,26 @@ protected void afterAlways(List<Throwable> errors) throws Throwable {
+    /**
+     * helper to randomly perform on <code>consumer</code> with <code>value</code>
+     */
+    public static <T> void maybeSet(Consumer<T> consumer, T value) {
```
nit: can both new helper methods go somewhere further down in ESTestCase? From the comments it seems someone tries to keep some order here, and this would be the setup / cleanup section. Also maybe both should be public? Although protected also works, we need it in tests that subclass this anyway.

areek replied:
Thanks for the suggestion, moved the helper methods to the right section

cbuescher commented:
@areek thanks, left one minor comment and I'm getting a few test failures (StringDistanceImplTests, TermSuggestionBuilderTests, CompletionSuggesterBuilderTests), some of which might already be fixed on the feature branch. 
Otherwise LGTM. Implementing parsing via innerFromXContent can be another PR, or did you want to add that here?

areek commented:
Thanks @cbuescher for looking, I think we should merge this in and implement parsing via innerFromXContent for CompletionSuggestionBuilder in a subsequent PR.
